robottelo/config/proxy.py

```diff
@@ -0,0 +1,238 @@
+
+        return version
+
+    def __capsule_hostname(self):
+        try:
+            instance_name = self.get('capsule.instance_name')
+            domain = self.get('capsule.domain')
+            return f"{instance_name}.{domain}"
+        except KeyError():
+            return None
+
+    def __ssh_client_command_timeout(self):
+        try:
+            timeout = self._get_from_configs('ssh_client.command_timeout')
+        except AttributeError:
+            timeout = 300
+        return timeout
+
+    def __ssh_client_connection_timeout(self):
+        try:
+            timeout = self._get_from_configs('ssh_client.connection_timeout')
+        except AttributeError:
+            timeout = 10
+        return timeout
+
+    def _dispatch_computed_value(self, key):
+        if key == "configure":
+            value = self._cached_function(lambda: None)
+        elif key == "server.get_credentials":
+            value = self._cached_function(self.__server_get_credentials)
+        elif key == "server.get_url":
+            value = self._cached_function(self.__server_get_url)
+        elif key == "server.get_pub_url":
+            value = self._cached_function(self.__server_get_pub_url)
+        elif key == "server.get_cert_rpm_url":
+            value = self._cached_function(self.__server_get_cert_rpm_url)
+        elif key == "server.version":
+            value = self.__server_version()
+        elif key == "capsule.hostname":
+            value = self.__capsule_hostname()
+        elif key == "ssh_client.command_timeout":
+            value = self.__ssh_client_command_timeout()
+        elif key == "ssh_client.connection_timeout":
+            value = self.__ssh_client_connection_timeout()
+        elif key == "configured":
+            value = True
+        elif key == "all_features":
+            value = self.__all_features()
+        else:
+            raise KeyError()
+        self._add_to_cache(key, value)
+        return value
+
+    def _get_from_configs(self, key):
+        for config_provider in self._configs:
+            try:
+                real_value = reduce(getattr, key.split('.'), config_provider)
+                logger.debug(
+                    f"obtained '{key}' from '{type(config_provider).__name__}' = {real_value}"
+                )
+                break
+            except AttributeError:
+                pass
+        else:
+            logger.debug(f"failed to find '{key}' in configuration")
+            msg = f"None of configuration providers has attribute '{key}'"
+            raise AttributeError(msg)
+        self._add_to_cache(key, real_value)
+        return real_value
+
+    def get(self, full_path):
+        try:
+            return self._from_cache(full_path)
+        except KeyError:
+            pass
+
+        try:
+            return self._dispatch_computed_value(full_path)
+        except KeyError:
+            pass
+
+        value = self._get_from_configs(full_path)
+        return value
```

Comment:
untested, but looks better imo
```suggestion
        return (
            self._from_cache.get (full_path) or 
            self._dispatch_computed_value.get(full_path) or 
            self._get_from_configs(full_path)
        )
```

Reply:
Don't forget that falsy values are completely normal and valid - `self._from_cache(full_path)` can return False, None, empty string, 0 (zero) or empty list. How would you differentiate between None-as-valid-value and None-as-could-not-find?

I could use sentinel value and direct comparison between it and whatever function returned, but at that point I just use `if` instead of `try/except`. Some people swear that `try/except` is not flow control, but in Python we believe in EAFP.